### ISSUE TITLE
A different approach to fields

### ISSUE THREAD

**Issue by treeowl** [kcsongor/generic-lens#136]
`Data.Generics.Product.Fields` forces `s field b -> t, t field a -> s` in a rather complicated way. But this fundep isn't even always desirable. For example, the `getConst` lens is quite reasonably type-changing.

A different approach:

1. Use (or crib from) `coercible-utils` to impose `Similar s t`, which means they both have the same type constructor and kind arguments.
2. Let the fact that `s field b -> Rep t` and `t field a -> Rep s` for each concrete type fall out of the constraint solving.

I think this way is much simpler, and it avoids unnecessary constraints; inference should be quite good, handling all non-phantom types.

If you don't want to do that in this package, let me know and I'll make my own for that.

**Comment by kcsongor** [comment]
There might be some subtleties I'm missing, but I think we already a have a similar lens:

https://github.com/kcsongor/generic-lens/blob/8e1fc7dcf444332c474fca17110d4bc554db08c8/generic-lens/src/Data/Generics/Product/Fields.hs#L112-L124
with the instance defined as
https://github.com/kcsongor/generic-lens/blob/8e1fc7dcf444332c474fca17110d4bc554db08c8/generic-lens/src/Data/Generics/Product/Fields.hs#L167-L168

https://github.com/kcsongor/generic-lens/blob/8e1fc7dcf444332c474fca17110d4bc554db08c8/generic-lens-core/src/Data/Generics/Product/Internal/Fields.hs#L44-L50

This uses

https://github.com/kcsongor/generic-lens/blob/8e1fc7dcf444332c474fca17110d4bc554db08c8/generic-lens-core/src/Data/Generics/Internal/Families/Changing.hs#L133-L137

This machinery looks similar to `Similar` at a first glance. Is this what you're after? If there's anything more we can do, I'm happy to add it to the package.

For concrete types this works out quite well, but in my experience, the fundeps come in really handy in a polymorphic context (like when a function has a `HasField` constraint, then without the fundeps, GHC will complain about ambiguous constraints).

**Comment by treeowl** [comment]
It's not quite the same, no. I'm talking about something like

```haskell
class (Similar s t, FindFieldType fld s ~ 'Just a, FindFieldType fld t ~ 'Just b) => Fielded fld s t a b where
  fielded :: Lens s t a b
```

This gives you `s -> a` and `t -> b`.
If you have `s ~ T y z`, then you immediately get `t ~ T _ _` and vice versa; I'm not sure if your `Field_` does the latter; it seems not to do the former.

**Comment by treeowl** [comment]
My version makes an a priori assumption over `HasField_` that the type is not void. This buys those two fundeps. I think that's a reasonable restriction. For a version offering only a `Traversal`, the restriction is best phrased a bit differently: at least one constructor must have a field by the given name. In that context, the restriction not only improves inference but also catches typos—you wouldn't usually want to get guaranteed-empty traversals if you misspell a field name.

**Comment by kcsongor** [comment]
The purpose of `UnifyHead` is exactly to infer from `s ~ T y z` that `t ~ T _ _`. Having `UnifyHead (T y z) t)` will give you `t ~ T _ _` after instance resolution (the first instance matches twice to take apart the application, then finally the second instance unifies the constructors). Looking at `Similar`, it seems to achieve the same thing in a slightly different way, using type families.

We run `UnifyHead` in both directions because we don't know which of `t` or `s` is more refined, i.e. which one will allow us to learn the shape of the other. 

Assuming that `UnifyHead` and `Similar` achieve the same thing (if they don't I'm happy to be corrected), the other thing you propose is writing the class declaration in the following way:

```haskell
class ... => Fielded fld s t a b | fld s -> a, fld t -> s where
```

in other words, we could perhaps add those fundeps to `HasField_`? Inference should be pretty good in concrete contexts. In fact `HasField_`s inference is already as good as the way more complex `HasField`'s in concrete contexts. It's polymorphic contexts that `HasField` is better in. Adding those two fundeps might bring them a little closer?
